src/lib/krb5/krb/get_etype_salt_s2kp.c

```diff
@@ -0,0 +1,118 @@
+/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+/* lib/krb5/krb/get_etype_salt_s2kp.c */
+/*
+ * Copyright (C) 2017 by the Massachusetts Institute of Technology.
+ * All rights reserved.
+ *
+ * Export of this software from the United States of America may
+ *   require a specific license from the United States Government.
+ *   It is the responsibility of any person or organization contemplating
+ *   export to obtain such a license before exporting.
+ *
+ * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
+ * distribute this software and its documentation for any purpose and
+ * without fee is hereby granted, provided that the above copyright
+ * notice appear in all copies and that both that copyright notice and
+ * this permission notice appear in supporting documentation, and that
+ * the name of M.I.T. not be used in advertising or publicity pertaining
+ * to distribution of the software without specific, written prior
+ * permission.  Furthermore if you modify this software you must label
+ * your software as modified software and not distribute it in such a
+ * fashion that it might be confused with the original M.I.T. software.
+ * M.I.T. makes no representations about the suitability of
+ * this software for any purpose.  It is provided "as is" without express
+ * or implied warranty.
+ */
+
+#include "k5-int.h"
+#include "init_creds_ctx.h"
+
+/*
+ * Send an AS_REQ to KDC and return etype, salt and s2kparams from AS_REP/ERROR
+ */
+krb5_error_code KRB5_CALLCONV
+krb5_get_etype_salt_s2kp(krb5_context context, krb5_principal principal,
+                         krb5_enctype *enctypes, krb5_enctype *enctype,
+                         krb5_data *salt, krb5_data *s2kparams)
+{
+    krb5_get_init_creds_opt *opt;
+    krb5_init_creds_context icc = NULL;
+    krb5_data reply = empty_data(), request = empty_data(), realm = empty_data();
+    krb5_error *error;
+    krb5_kdc_rep *asrep;
+    krb5_pa_data **padata;
+    unsigned int flags;
+    int master = 0, nenctypes = 0;
+    krb5_error_code retval = 0;
+
+    retval = krb5_get_init_creds_opt_alloc(context, &opt);
+    if (retval)
+        goto cleanup;
+    if (enctypes) {
+        nenctypes = k5_count_etypes(enctypes);
+        if (nenctypes)
+            krb5_get_init_creds_opt_set_etype_list(opt, enctypes, nenctypes);
+    }
+
+    retval = krb5_init_creds_init(context, principal, NULL, NULL, 0, opt, &icc);
+    if (retval)
+        goto cleanup;
+    retval = krb5_init_creds_step(context, icc, &reply, &request, &realm,
+                                  &flags);
+    if (retval)
+        goto cleanup;
+    if (flags != KRB5_INIT_CREDS_STEP_FLAG_CONTINUE) {
+        retval = KRB5KRB_AP_ERR_MSG_TYPE;
+        goto cleanup;
+    }
+    retval = krb5_sendto_kdc(context, &request, &realm, &reply, &master, 1);
+    if (retval)
+        goto cleanup;
+
+    if (krb5_is_krb_error(&reply)) {
```

Comment:
I would like to see a helper function to extract padata from reply.  Currently the memory management appears to leak in a number of error cases, and I think using a helper function will help.

I would also be okay with keeping the krb5_is_krb_error() conditional in this function, and making two separate helpers for extracting padata from KRB-ERROR and AS-REP.  Whichever approach makes the code come out simpler is fine.